lib/std/Build.zig

```diff
@@ -1867,6 +1867,33 @@ pub fn dependency(b: *Build, name: []const u8, args: anytype) *Dependency {
     unreachable; // Bad @dependencies source
 }
 
+pub fn dependencyFromBuildZig(
+    b: *std.Build,
+    /// The struct that corresponds to the build.zig of the dependency (usually
+    /// obtained by direct `@import` of the build.zig, or `@This` if called from
+    /// a function declared inside the build.zig itself).
+    comptime build_zig: type,
+    args: anytype,
+) *std.Build.Dependency {
+    const build_runner = @import("root");
+    const deps = build_runner.dependencies;
+
+    find_dep: {
+        const pkg, const pkg_hash = inline for (@typeInfo(deps.packages).Struct.decls) |decl| {
+            const pkg_hash = decl.name;
+            const pkg = @field(deps.packages, pkg_hash);
+            if (@hasDecl(pkg, "build_zig") and pkg.build_zig == build_zig) break .{ pkg, pkg_hash };
+        } else break :find_dep;
+        const dep_name = for (b.available_deps) |dep| {
```

Comment:
I should add that there's no need to check for `@hasDecl(pkg, "available")` because if we get this far, we must have `@import`ed the dependency and know for sure that the package is already fetched.